src/basic/ml/FStar_Util.ml

```diff
@@ -894,58 +895,84 @@ let write_hints (filename: string) (hints: hints_db): unit =
                 `List (List.map (fun s -> `String s) strings));
-            `Int (Z.to_int query_elapsed_time)
+            `Int (Z.to_int query_elapsed_time);
+            `String (match hash with | Some(h) -> h | _ -> "")
           ]
     ) hints.hints)
   ] in
   Yojson.Safe.pretty_to_channel (open_out_bin filename) json
 
 let read_hints (filename: string): hints_db option =
-    try
+  try
     let chan = open_in filename in
     let json = Yojson.Safe.from_channel chan in
     close_in chan;
-    Some (match json with
-    | `List [
-        `String module_digest;
-        `List hints
-      ] ->
-        {
-          module_digest;
-          hints = List.map (function
-            | `Null -> None
-            | `List [
-		`String hint_name;
-		`Int hint_index;
-                `Int fuel;
-                `Int ifuel;
-                unsat_core;
-                `Int query_elapsed_time
-              ] ->
-                Some {
-		  hint_name;
-		  hint_index = Z.of_int hint_index;
-                  fuel = Z.of_int fuel;
-                  ifuel = Z.of_int ifuel;
-                  unsat_core = begin match unsat_core with
-                    | `Null ->
-                        None
-                    | `List strings ->
-                        Some (List.map (function
-                          | `String s -> s
-                          | _ -> raise Exit
-                        ) strings)
-                    |  _ ->
-                        raise Exit
-                  end;
-                  query_elapsed_time = Z.of_int query_elapsed_time
-                }
-              | _ ->
-                 raise Exit
-          ) hints
-        }
-    | _ ->
-        raise Exit
+    Some (
+        match json with
+        | `List [
+            `String module_digest;
+            `List hints
+          ] -> {
+            module_digest;
+            hints = List.map (function
+                        | `Null -> None
+                        | `List [ `String hint_name;
+		                          `Int hint_index;
+                                  `Int fuel;
+                                  `Int ifuel;
+                                  unsat_core;
+                                  `Int query_elapsed_time ] ->
+                           Some {
+		                       hint_name;
+		                       hint_index = Z.of_int hint_index;
+                               fuel = Z.of_int fuel;
+                               ifuel = Z.of_int ifuel;
+                               unsat_core = begin match unsat_core with
+                                            | `Null ->
+                                               None
+                                            | `List strings ->
+                                               Some (List.map (function
+                                                         | `String s -> s
+                                                         | _ -> raise Exit
```

Comment:
One problem with using Exceptions for control flow like this is that it yields very slow Javascript when we extract. But I think this code doesn't get run a lot, right?

Reply:
Yep, that runs once while reading the hints file.